internal/telemetry/telemetry.go

```diff
@@ -43,26 +45,59 @@ type Event struct {
 	Properties map[string]interface{} `json:"properties"`
 }
 
+type telemetryContextKey struct{}
+
+type telemetryContextValue struct {
+	startTime time.Time
+}
+
+func NewContext() context.Context {
+	return context.WithValue(context.Background(), contextKey, telemetryContextValue{
+		startTime: time.Now(),
+	})
+}
+
+func valueFromContext(ctx context.Context) (telemetryContextValue, bool) {
+	value, ok := ctx.Value(contextKey).(telemetryContextValue)
+	return value, ok
+}
+
 func TrackCommand(cmd *cobra.Command) {
 	if !config.TelemetryEnabled() {
 		return
 	}
 	track(cmd)
 }
 
-func track(cmd *cobra.Command) {
+func newEvent(cmd *cobra.Command) Event {
 	now := time.Now()
 	cmdPath := cmd.CommandPath()
 	command := strings.ReplaceAll(cmdPath, " ", "-")
+
+	ctxValue, found := valueFromContext(cmd.Context())
+	var duration time.Duration
+	if found {
+		duration = now.Sub(ctxValue.startTime)
+	}
```

Comment:
if `found == false` the duration will default to 0 I think.
Maybe in this case we should return an error and the `track` function below to handle that error (log and return)?
(so the signature of this function will be `func newEvent(cmd *cobra.Command) (*Event, error)`)

Reply:
Telemetry is meant to ignore errors completely so commands are never affected by telemetry, also I don't want to not track an event just because it did not use context correctly, chances are if we are receiving 0 duration context was not carried out to that specific command which would be about fixing it instead of not tracking.

Reply:
`track` function is only logging errors, not affecting the commands, that's what I was also thinking when sugested "and the track function below to handle that error"
Even if capturing the event is more important than, in this case, the correctness of its duration, we should not simply hide a possible mishandling of the context. At least we should log that as an error (like we do in the track function)

Reply:
@tibulca I'm now calling logError, let me know what you think